File: backend/src/mesh/Packet.ts
export enum PacketType {
  HELLO = 'HELLO',
  IDENTITY = 'IDENTITY',
  SESSION_INIT = 'SESSION_INIT',
  SESSION_ACK = 'SESSION_ACK',
  MESSAGE = 'MESSAGE',
  MESSAGE_ACK = 'MESSAGE_ACK',
  FILE_OFFER = 'FILE_OFFER',
  FILE_CHUNK = 'FILE_CHUNK',
  FILE_ACK = 'FILE_ACK',
  ROUTE_DISCOVERY = 'ROUTE_DISCOVERY',
  ROUTE_UPDATE = 'ROUTE_UPDATE',
  ERROR = 'ERROR',
  EMERGENCY_BROADCAST = 'EMERGENCY_BROADCAST',
}

export interface Packet<T = any> {
  packetId: string;
  sourceId: string;
  destinationId: string | null; // null for broadcast
  type: PacketType;
  ttl: number;                  // Time to live (e.g. 5 hops max)
  sequence: number;             // Monotonically increasing sequence number
  timestamp: number;
  payload: T;
  signature?: string;           // Optional cryptographic signature
  routeTrace?: string[];        // Relay nodes this packet traversed
}

export class PacketFactory {
  private static sequenceCounter = 0;

  public static create<T>(
    sourceId: string,
    destinationId: string | null,
    type: PacketType,
    payload: T,
    ttl = 5,
    signature?: string
  ): Packet<T> {
    this.sequenceCounter++;
    return {
      packetId: `pkt-${Date.now()}-${Math.random().toString(36).substring(2, 9)}`,
      sourceId,
      destinationId,
      type,
      ttl,
      sequence: this.sequenceCounter,
      timestamp: Date.now(),
      payload,
      signature,
      routeTrace: [sourceId],
    };
  }
}

File: android/domain/model/Packet.kt
package com.rescom.domain.model

enum class PacketType {
    HELLO,
    IDENTITY,
    SESSION_INIT,
    SESSION_ACK,
    MESSAGE,
    MESSAGE_ACK,
    FILE_OFFER,
    FILE_CHUNK,
    FILE_ACK,
    ROUTE_DISCOVERY,
    ROUTE_UPDATE,
    EMERGENCY_BROADCAST,
    ERROR
}

data class Packet(
    val packetId: String,
    val sourceId: String,
    val destinationId: String?,
    val type: PacketType,
    val ttl: Int = 5,
    val sequence: Long,
    val timestamp: Long = System.currentTimeMillis(),
    val payload: ByteArray,
    val signature: String? = null,
    val routeTrace: List<String> = emptyList()
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Packet
        return packetId == other.packetId
    }

    override fun hashCode(): Int {
        return packetId.hashCode()
    }
}
